sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,162 @@ class GpuDynamicPartitionDataSingleWriter(
     assert(!isBucketed)
 
     // We have an entire batch that is sorted, so we need to split it up by key
-    var needToCloseBatch = true
-    var partitionColumns: Table = null
-    var distinctKeys: Table = null
-    var outputColumns: Table = null
-    var splits: Array[ContiguousTable] = null
-    var cbKeys: ColumnarBatch = null
     val maxRecordsPerFile = description.maxRecordsPerFile
-
-    try {
-      partitionColumns = getPartitionColumns(cb)
-      val partDataTypes = description.partitionColumns.map(_.dataType).toArray
-      distinctKeys = distinctAndSort(partitionColumns)
-      val partitionIndexes = splitIndexes(partitionColumns, distinctKeys)
-      partitionColumns.close()
-      partitionColumns = null
+    val partDataTypes = description.partitionColumns.map(_.dataType).toArray
+    val outDataTypes = description.dataColumns.map(_.dataType).toArray
+    val (splits, cbKeys) = withResource(cb) { cb =>
+      val (partitionIndexes, cbKeys) = withResource(getPartitionColumns(cb)) { partitionColumns =>
+        withResource(distinctAndSort(partitionColumns)) { distinctKeys =>
+          val partitionIndexes = splitIndexes(partitionColumns, distinctKeys)
+          val cbKeys = copyToHostAsBatch(distinctKeys, partDataTypes)
+          (partitionIndexes, cbKeys)
+        }
+      }
 
       // split the original data on the indexes
-      outputColumns = getOutputColumns(cb)
-      val outDataTypes = description.dataColumns.map(_.dataType).toArray
-      cb.close()
-      needToCloseBatch = false
-      splits = outputColumns.contiguousSplit(partitionIndexes: _*)
-      outputColumns.close()
-      outputColumns = null
-
-      cbKeys = copyToHostAsBatch(distinctKeys, partDataTypes)
-      distinctKeys.close()
-      distinctKeys = null
-
-      // Use the existing code to convert each row into a path. It would be nice to do this on
-      // the GPU, but the data should be small and there are things we cannot easily support
-      // on the GPU right now
-      import scala.collection.JavaConverters._
-      val paths = cbKeys.rowIterator().asScala.map(getPartitionPath)
-
-      paths.toArray.zip(splits).foreach(combined => {
-        val table = combined._2.getTable
-        val batch = GpuColumnVector.from(table, outDataTypes)
-        val partPath = combined._1
-
-        // If fall back from for `GpuDynamicPartitionDataConcurrentWriter`, we should get the
-        // saved status
-        var savedStatus: Option[WriterStatusWithCaches] = None
-
-        if (currentPartPath != partPath) {
-          val previousPartPath = currentPartPath
-          currentPartPath = partPath
+      val splits = withResource(getOutputColumns(cb)) { outputColumns =>
```

Comment:
`cbKeys` is closeable, so need to close it if any exception.
```
val splits = closeOnExcept(cbKeys) { _ =>
  withResource(getOutputColumns(cb)) { outputColumns =>
    .....
  }
}
```

Reply:
Great catch!

Reply:
Updated.